File: fitness_engine/models/assessment.py
"""
Assessment result data models — output of the assessment module.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Optional

# Phase-6 cleanup: hoisted from inside BodyComposition.__post_init__ and
# MuscularPotential.__post_init__ (was a deferred import for no reason).
import math

# Phase-6 cleanup: shared JSON-serializer replaces the per-class ``_convert``
# helper that was duplicated across assessment / nutrition models.
from ..utils.serialize import convert_for_json


class BodyFatMethod(str, Enum):
    """Method used to compute body fat %."""
    USER_PROVIDED = "user_provided"
    NAVY = "navy"                  # US Navy circumference (Hodgdon & Beckett 1984)
    BMI_JACKSON = "bmi_jackson"   # Jackson et al. 2002 (BMI-based)
    CUN_BAE = "cun_bae"            # Gomez-Ambrosi 2012


class BodyFatCategory(str, Enum):
    ESSENTIAL = "essential"
    ATHLETE = "athlete"
    FITNESS = "fitness"
    ACCEPTABLE = "acceptable"
    OBESITY = "obesity"


class BMICategory(str, Enum):
    UNDERWEIGHT = "underweight"
    NORMAL = "normal"
    OVERWEIGHT = "overweight"
    OBESE = "obese"


class HealthRiskLevel(str, Enum):
    LOW = "low"
    MODERATE = "moderate"
    HIGH = "high"
    VERY_HIGH = "very_high"


class ABSIRiskLevel(str, Enum):
    LOW = "low"
    BELOW_AVERAGE = "below_average"
    AVERAGE = "average"
    ABOVE_AVERAGE = "above_average"
    HIGH = "high"


class RecommendedStrategy(str, Enum):
    CUT = "cut"
    BULK = "bulk"
    RECOMP = "recomp"
    MAINTENANCE = "maintenance"
    HABIT_CHANGE_FIRST = "habit_change_first"


@dataclass
class BodyComposition:
    """Body composition assessment results."""
    body_fat_pct: float
    body_fat_method: BodyFatMethod
    body_fat_category: BodyFatCategory
    lean_body_mass_kg: float
    fat_mass_kg: float
    bmi: float
    bmi_category: BMICategory
    ffmi: float                              # Fat-Free Mass Index
    normalized_ffmi: float                   # height-adjusted (Kouri 1995)
    target_weight_at_target_bf_kg: Optional[float] = None
    notes: list[str] = field(default_factory=list)

    def __post_init__(self):
        """Tier 3.32 fix: validate output ranges to catch bugs that would
        silently propagate impossible values (e.g. BF%=250, FFMI=-5)."""
        if math.isnan(self.body_fat_pct) or not (0 <= self.body_fat_pct <= 100):
            raise ValueError(
                f"BodyComposition.body_fat_pct must be in [0, 100]; got {self.body_fat_pct}"
            )
        if math.isnan(self.bmi) or self.bmi <= 0:
            raise ValueError(f"BodyComposition.bmi must be positive; got {self.bmi}")
        if math.isnan(self.ffmi) or self.ffmi < 0:
            raise ValueError(f"BodyComposition.ffmi must be non-negative; got {self.ffmi}")
        if math.isnan(self.lean_body_mass_kg) or self.lean_body_mass_kg < 0:
            raise ValueError(
                f"BodyComposition.lean_body_mass_kg must be non-negative; "
                f"got {self.lean_body_mass_kg}"
            )
        if math.isnan(self.fat_mass_kg) or self.fat_mass_kg < 0:
            raise ValueError(
                f"BodyComposition.fat_mass_kg must be non-negative; got {self.fat_mass_kg}"
            )


@dataclass
class HealthRiskAssessment:
    """Health risk metrics based on body measurements."""
    whr: Optional[float] = None                       # waist-to-hip
    whr_risk: Optional[HealthRiskLevel] = None
    whtr: Optional[float] = None                      # waist-to-height
    whtr_risk: Optional[HealthRiskLevel] = None
    absi: Optional[float] = None                      # A Body Shape Index
    absi_z_score: Optional[float] = None
    absi_risk: Optional[ABSIRiskLevel] = None
    ibw_devine_kg: Optional[float] = None
    ibw_robinson_kg: Optional[float] = None
    ibw_miller_kg: Optional[float] = None
    ibw_hamwi_kg: Optional[float] = None
    overall_risk: HealthRiskLevel = HealthRiskLevel.LOW
    risk_factors: list[str] = field(default_factory=list)
    notes: list[str] = field(default_factory=list)


@dataclass
class MuscularPotential:
    """Assessment of muscular development relative to genetic potential."""
    current_ffmi: float
    current_normalized_ffmi: float
    # Phase-6 fix: these literals mirror the canonical constants in
    # assessment.muscular_potential (FFMI_NATURAL_COMMON/ATTAINABLE/LIKELY_MAX).
    # We can't import them at module top-level here because muscular_potential
    # imports MuscularPotential from this module (circular). The values are kept
    # in sync via the cross-reference comments on both sides; floats are
    # immutable so there is no mutable-default issue.
    natural_ceiling_ffmi: float = 25.0              # == FFMI_NATURAL_COMMON (Kouri 1995)
    attainable_ceiling_ffmi: float = 27.3            # == FFMI_NATURAL_ATTAINABLE (Mr. America 1939-1953)
    likely_max_ffmi: float = 28.0                    # == FFMI_NATURAL_LIKELY_MAX (RippedBody editorial)
    berkhan_stage_max_kg: Optional[float] = None     # height_cm - 100
    ffmi_to_ceiling_pct: float = 0.0                 # current / natural ceiling
    headroom_kg: float = 0.0                         # FFM kg remaining to ceiling
    expected_monthly_muscle_gain_kg: float = 0.0     # by training status
    # Phase-6 fix: flag when normalized FFMI exceeds the natural ceiling
    # (PED users, genetic outliers) so ffmi_to_ceiling_pct can be clamped to 100
    # without losing the signal that the user is over-ceiling.
    is_above_ceiling: bool = False
    notes: list[str] = field(default_factory=list)

    def __post_init__(self):
        """Tier 3.32 fix: validate output ranges."""
        if math.isnan(self.current_ffmi) or self.current_ffmi < 0:
            raise ValueError(f"MuscularPotential.current_ffmi must be non-negative; got {self.current_ffmi}")
        if math.isnan(self.current_normalized_ffmi) or self.current_normalized_ffmi < 0:
            raise ValueError(
                f"MuscularPotential.current_normalized_ffmi must be non-negative; "
                f"got {self.current_normalized_ffmi}"
            )
        # Phase-6 fix: enforce all numeric fields are non-negative AND non-NaN
        # (was inconsistent — only first two fields had NaN checks).
        for fname, fval in (
            ("expected_monthly_muscle_gain_kg", self.expected_monthly_muscle_gain_kg),
            ("headroom_kg", self.headroom_kg),
            ("ffmi_to_ceiling_pct", self.ffmi_to_ceiling_pct),
        ):
            if math.isnan(fval) or fval < 0:
                raise ValueError(
                    f"MuscularPotential.{fname} must be non-negative and non-NaN; "
                    f"got {fval}"
                )
        # Phase-6 fix: ffmi_to_ceiling_pct is documented as ≤100; enforce both
        # bounds (was previously only checking the lower bound).
        if self.ffmi_to_ceiling_pct > 100:
            raise ValueError(
                f"MuscularPotential.ffmi_to_ceiling_pct must be ≤100; "
                f"got {self.ffmi_to_ceiling_pct}"
            )


@dataclass
class AssessmentResult:
    """Top-level assessment output combining all sub-assessments."""
    body_composition: BodyComposition
    health_risk: HealthRiskAssessment
    muscular_potential: MuscularPotential
    recommended_strategy: RecommendedStrategy
    strategy_rationale: str
    summary: str                                     # human-readable summary

    def to_dict(self) -> dict:
        """JSON-serializable view."""
        # Phase-6 cleanup: uses the shared ``convert_for_json`` helper from
        # ``utils.serialize`` (was a duplicated local ``_convert`` function).
        return convert_for_json(self)


__all__ = [
    "BodyFatMethod",
    "BodyFatCategory",
    "BMICategory",
    "HealthRiskLevel",
    "ABSIRiskLevel",
    "RecommendedStrategy",
    "BodyComposition",
    "HealthRiskAssessment",
    "MuscularPotential",
    "AssessmentResult",
]

File: download/fitn/fitn_engine/lib/src/models/assessment.dart
/// Assessment output models. See spec §9.1 (output tree).
library;

import 'enums.dart';

class BodyComposition {
  BodyComposition({
    required this.bodyFatPct,
    required this.bodyFatMethod,
    required this.bodyFatCategory,
    required this.leanBodyMassKg,
    required this.fatMassKg,
    required this.bmi,
    required this.bmiCategory,
    required this.ffmi,
    required this.normalizedFfmi,
    required this.targetWeightsKg,
    required this.notes,
  });

  final double bodyFatPct;
  final BodyFatMethod bodyFatMethod;
  final BodyFatCategory bodyFatCategory;
  final double leanBodyMassKg;
  final double fatMassKg;
  final double bmi;
  final BMICategory bmiCategory;
  final double ffmi;
  final double normalizedFfmi;
  final Map<String, double> targetWeightsKg; // {athletic, fitness, acceptable, hormonal_floor}
  final List<String> notes;

  Map<String, dynamic> toJson() => {
        'body_fat_pct': bodyFatPct,
        'body_fat_method': bodyFatMethod.toJson(),
        'body_fat_category': bodyFatCategory.toJson(),
        'lean_body_mass_kg': leanBodyMassKg,
        'fat_mass_kg': fatMassKg,
        'bmi': bmi,
        'bmi_category': bmiCategory.toJson(),
        'ffmi': ffmi,
        'normalized_ffmi': normalizedFfmi,
        'target_weights_kg': targetWeightsKg,
        'notes': notes,
      };

  factory BodyComposition.fromJson(Map<String, dynamic> json) {
    return BodyComposition(
      bodyFatPct: (json['body_fat_pct'] as num).toDouble(),
      bodyFatMethod:
          BodyFatMethodJson.fromJson(json['body_fat_method'] as String),
      bodyFatCategory:
          BodyFatCategoryJson.fromJson(json['body_fat_category'] as String),
      leanBodyMassKg: (json['lean_body_mass_kg'] as num).toDouble(),
      fatMassKg: (json['fat_mass_kg'] as num).toDouble(),
      bmi: (json['bmi'] as num).toDouble(),
      bmiCategory: BMICategoryJson.fromJson(json['bmi_category'] as String),
      ffmi: (json['ffmi'] as num).toDouble(),
      normalizedFfmi: (json['normalized_ffmi'] as num).toDouble(),
      targetWeightsKg: (json['target_weights_kg'] as Map).map(
          (k, v) => MapEntry(k as String, (v as num).toDouble())),
      notes: (json['notes'] as List).cast<String>(),
    );
  }
}

class HealthRiskAssessment {
  HealthRiskAssessment({
    required this.whr,
    required this.whrRisk,
    required this.whtr,
    required this.whtrRisk,
    required this.absi,
    required this.absiZScore,
    required this.absiRisk,
    required this.ibwDevineKg,
    required this.ibwRobinsonKg,
    required this.ibwMillerKg,
    required this.ibwHamwiKg,
    required this.overallRisk,
    required this.riskFactors,
    required this.notes,
  });

  final double? whr;
  final HealthRiskLevel? whrRisk;
  final double? whtr;
  final HealthRiskLevel? whtrRisk;
  final double? absi;
  final double? absiZScore;
  final ABSIRiskLevel? absiRisk;
  final double? ibwDevineKg;
  final double? ibwRobinsonKg;
  final double? ibwMillerKg;
  final double? ibwHamwiKg;
  final HealthRiskLevel overallRisk;
  final List<String> riskFactors;
  final List<String> notes;

  Map<String, dynamic> toJson() => {
        'whr': whr,
        'whr_risk': whrRisk?.toJson(),
        'whtr': whtr,
        'whtr_risk': whtrRisk?.toJson(),
        'absi': absi,
        'absi_z_score': absiZScore,
        'absi_risk': absiRisk?.toJson(),
        'ibw_devine_kg': ibwDevineKg,
        'ibw_robinson_kg': ibwRobinsonKg,
        'ibw_miller_kg': ibwMillerKg,
        'ibw_hamwi_kg': ibwHamwiKg,
        'overall_risk': overallRisk.toJson(),
        'risk_factors': riskFactors,
        'notes': notes,
      };

  factory HealthRiskAssessment.fromJson(Map<String, dynamic> json) {
    return HealthRiskAssessment(
      whr: (json['whr'] as num?)?.toDouble(),
      whrRisk: json['whr_risk'] != null
          ? HealthRiskLevelJson.fromJson(json['whr_risk'] as String)
          : null,
      whtr: (json['whtr'] as num?)?.toDouble(),
      whtrRisk: json['whtr_risk'] != null
          ? HealthRiskLevelJson.fromJson(json['whtr_risk'] as String)
          : null,
      absi: (json['absi'] as num?)?.toDouble(),
      absiZScore: (json['absi_z_score'] as num?)?.toDouble(),
      absiRisk: json['absi_risk'] != null
          ? ABSIRiskLevelJson.fromJson(json['absi_risk'] as String)
          : null,
      ibwDevineKg: (json['ibw_devine_kg'] as num?)?.toDouble(),
      ibwRobinsonKg: (json['ibw_robinson_kg'] as num?)?.toDouble(),
      ibwMillerKg: (json['ibw_miller_kg'] as num?)?.toDouble(),
      ibwHamwiKg: (json['ibw_hamwi_kg'] as num?)?.toDouble(),
      overallRisk:
          HealthRiskLevelJson.fromJson(json['overall_risk'] as String),
      riskFactors: (json['risk_factors'] as List).cast<String>(),
      notes: (json['notes'] as List).cast<String>(),
    );
  }
}

class MuscularPotential {
  MuscularPotential({
    required this.currentFfmi,
    required this.currentNormalizedFfmi,
    required this.naturalCeilingFfmi,
    required this.attainableCeilingFfmi,
    required this.likelyMaxFfmi,
    required this.berkhanStageMaxKg,
    required this.ffmiToCeilingPct,
    required this.headroomKg,
    required this.expectedMonthlyMuscleGainKg,
    required this.isAboveCeiling,
    required this.notes,
  });

  final double currentFfmi;
  final double currentNormalizedFfmi;
  final double naturalCeilingFfmi;
  final double attainableCeilingFfmi;
  final double likelyMaxFfmi;
  final double? berkhanStageMaxKg;
  final double ffmiToCeilingPct;
  final double headroomKg;
  final double expectedMonthlyMuscleGainKg;
  final bool isAboveCeiling;
  final List<String> notes;

  Map<String, dynamic> toJson() => {
        'current_ffmi': currentFfmi,
        'current_normalized_ffmi': currentNormalizedFfmi,
        'natural_ceiling_ffmi': naturalCeilingFfmi,
        'attainable_ceiling_ffmi': attainableCeilingFfmi,
        'likely_max_ffmi': likelyMaxFfmi,
        'berkhan_stage_max_kg': berkhanStageMaxKg,
        'ffmi_to_ceiling_pct': ffmiToCeilingPct,
        'headroom_kg': headroomKg,
        'expected_monthly_muscle_gain_kg': expectedMonthlyMuscleGainKg,
        'is_above_ceiling': isAboveCeiling,
        'notes': notes,
      };

  factory MuscularPotential.fromJson(Map<String, dynamic> json) {
    return MuscularPotential(
      currentFfmi: (json['current_ffmi'] as num).toDouble(),
      currentNormalizedFfmi:
          (json['current_normalized_ffmi'] as num).toDouble(),
      naturalCeilingFfmi:
          (json['natural_ceiling_ffmi'] as num).toDouble(),
      attainableCeilingFfmi:
          (json['attainable_ceiling_ffmi'] as num).toDouble(),
      likelyMaxFfmi: (json['likely_max_ffmi'] as num).toDouble(),
      berkhanStageMaxKg: (json['berkhan_stage_max_kg'] as num?)?.toDouble(),
      ffmiToCeilingPct:
          (json['ffmi_to_ceiling_pct'] as num).toDouble(),
      headroomKg: (json['headroom_kg'] as num).toDouble(),
      expectedMonthlyMuscleGainKg:
          (json['expected_monthly_muscle_gain_kg'] as num).toDouble(),
      isAboveCeiling: json['is_above_ceiling'] as bool,
      notes: (json['notes'] as List).cast<String>(),
    );
  }
}

class AssessmentResult {
  AssessmentResult({
    required this.bodyComposition,
    required this.healthRisk,
    required this.muscularPotential,
    required this.recommendedStrategy,
    required this.strategyRationale,
    required this.summary,
    required this.isPartial,
    required this.errors,
  });

  final BodyComposition? bodyComposition;
  final HealthRiskAssessment? healthRisk;
  final MuscularPotential? muscularPotential;
  final RecommendedStrategy recommendedStrategy;
  final String strategyRationale;
  final String summary;
  final bool isPartial;
  final List<String> errors;

  Map<String, dynamic> toJson() => {
        'body_composition': bodyComposition?.toJson(),
        'health_risk': healthRisk?.toJson(),
        'muscular_potential': muscularPotential?.toJson(),
        'recommended_strategy': recommendedStrategy.toJson(),
        'strategy_rationale': strategyRationale,
        'summary': summary,
        'is_partial': isPartial,
        'errors': errors,
      };

  factory AssessmentResult.fromJson(Map<String, dynamic> json) {
    return AssessmentResult(
      bodyComposition: json['body_composition'] != null
          ? BodyComposition.fromJson(
              json['body_composition'] as Map<String, dynamic>)
          : null,
      healthRisk: json['health_risk'] != null
          ? HealthRiskAssessment.fromJson(
              json['health_risk'] as Map<String, dynamic>)
          : null,
      muscularPotential: json['muscular_potential'] != null
          ? MuscularPotential.fromJson(
              json['muscular_potential'] as Map<String, dynamic>)
          : null,
      recommendedStrategy: RecommendedStrategyJson.fromJson(
          json['recommended_strategy'] as String),
      strategyRationale: json['strategy_rationale'] as String,
      summary: json['summary'] as String,
      isPartial: json['is_partial'] as bool,
      errors: (json['errors'] as List).cast<String>(),
    );
  }
}
